Source/Scene/ModelExperimental/MetadataPipelineStage.js

```diff
@@ -66,6 +91,22 @@ MetadataPipelineStage.process = function (
   processPropertyTextures(renderResources, structuralMetadata);
 };
 
+function declareMetadataClassStructs(shaderBuilder) {
+  for (const metadataType of MetadataPipelineStage.METADATA_TYPES) {
+    const structName = `${metadataType}MetadataClass`;
+    const structIdVs = `${structName}VS`;
+    const structIdFs = `${structName}FS`;
+    // Declare the struct in both vertex and fragment shaders
+    shaderBuilder.addStruct(structIdVs, structName, ShaderDestination.VERTEX);
+    shaderBuilder.addStruct(structIdFs, structName, ShaderDestination.FRAGMENT);
+    // Add fields
+    for (const { shaderName } of MetadataPipelineStage.METADATACLASS_FIELDS) {
```

Comment:
@ggetz or @sanjeetsuhag I know we enabled ES6 syntax, but I also know that `for-of loops` is always the most controversial feature performance-wise. Obviously we want to be careful in code that runs every frame, but are they okay in cases like this that run less frequently?

For the record, this code (and anything in these pipeline stages) here runs:
* once when the model first loads
* on an as-needed basis whenever larger changes are needed to the shader/other rendering resources
* Though multiply this by "for every tile in the tileset"

Furthermore, what's our policy on reformatting old code? (see also https://github.com/CesiumGS/cesium/issues/10486). I'd love to go back and modernize this `ModelExperimental` code like this PR does, but I know we don't like to fragment the codebase.

Reply:
But the most important question to answer for the sake of this PR is "is this use of ES6 okay, or should we revert to ES5 syntax so we don't fragment the codebase"

Reply:
Given that this is rendering resource related code and that it may be run repeatedly (not per-frame but still), my gut reaction is to defer to simpler, faster code through ES5.

Reply:
Based off the [coding guide](https://github.com/CesiumGS/cesium/blob/main/Documentation/Contributors/CodingGuide/README.md#basic-code-construction), we should at a minimum abide by this:

> Modern language features may provide handy shortcuts and cleaner syntax, but they should be used with consideration for their performance implications, especially in code that is invoked per-frame.

If this is not high-touch area for performance, then the for-of loop should be OK. Otherwise, no.

Something else to consider when using these new features is readability. Code is read far more than it is written, so new syntax should strive to make the code _more_ readable, not less.

While destructuring assignment _can_ conform to those two rules, l In this case where its used in the for loop clause:

````js
for (const { shaderName } of MetadataPipelineStage.METADATACLASS_FIELDS) {
````
is less readable than
````js
for (const field of MetadataPipelineStage.METADATACLASS_FIELDS) {
const shaderName = field.shaderName;
````

I would also be wary of using a lot of destructuring on undocumented `options` parameters getting passed around to helper functions, as it's easy to loose track of what that object actually is.

For fat arrow notation, I think it's common enough that readability isn't an issue, and gives us the benefit of keeping the same local variable scopes. IMO they are a net positive, but consider if we want to enforce any style guides around them such as https://eslint.org/docs/latest/rules/arrow-body-style.

In terms of fragmenting the code base, it would be great to not do so where we can, but that may not be possible in cases where there isn't an eslint rule which can automatically fix all instances in the codebase. To ensure consistency then, guidelines we decide on should should be noted in the coding guide or enforced through an eslint rule so the decision can be standardized.

Reply:
Thanks for the feedback! I just pushed 2 changes in this file:

**1. Rewrote ES6 for-of loops as ES5 for loops**
As @ptrgags and @sanjeetsuhag noted, this is not per-frame rendering code, but it could still run fairly frequently (per-tile). And in this particular case, the for-of syntax actually wasn't much more readable, so the faster ES5 loop makes sense.

**2. Reverted nested destructuring in `addValueTransformUniforms`**
@ggetz I agree that since the options object is not documented, nested destructuring like the following code could eventually lead to unclear error messages if the options object is changed in the future:
```javascript
// BAD?
const {
  renderResources: { shaderBuilder, uniformMap },
  ...
} = options;
```

I still kept smaller destructuring assignments like this one, for conciseness:
```javascript
// GOOD
const { shaderBuilder, uniformMap } = options.renderResources;
```

Let me know if this is still an issue. My understanding is that with the above line, any errors should be the same as in the more repetitive syntax:
```javascript
// REPETITIVE
const renderResources = options.renderResources;
const shaderBuilder = renderResources.shaderBuilder;
const uniformMap = renderResources.uniformMap;
```